flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/rules/logical/FlinkBushyJoinReorderRule.java

```diff
@@ -0,0 +1,620 @@
+                }
+                for (JoinPlan otherSidePlan : otherSideCandidates) {
+                    Optional<JoinPlan> newJoinPlan =
+                            buildJoin(
+                                    relBuilder, oneSidePlan, otherSidePlan, multiJoin, isOuterJoin);
+                    if (newJoinPlan.isPresent()) {
+                        JoinPlan existingPlan = nextLevel.get(newJoinPlan.get().factorIds);
+                        // check if it's the first plan for the factor set, or it's a better plan
+                        // than the existing one due to lower cost.
+                        if (existingPlan == null
+                                || newJoinPlan.get().betterThan(existingPlan, mq)) {
+                            nextLevel.put(newJoinPlan.get().factorIds, newJoinPlan.get());
+                        }
+                    }
+                }
+            }
+            k += 1;
+        }
+        return nextLevel;
+    }
+
+    private static Optional<JoinPlan> buildJoin(
+            RelBuilder relBuilder,
+            JoinPlan oneSidePlan,
+            JoinPlan otherSidePlan,
+            LoptMultiJoin multiJoin,
+            boolean isOuterJoin) {
+        // intersect, should not join two overlapping factor sets.
+        Set<Integer> resSet = new HashSet<>(oneSidePlan.factorIds);
+        resSet.retainAll(otherSidePlan.factorIds);
+        if (!resSet.isEmpty()) {
+            return Optional.empty();
+        }
+
+        Optional<Tuple2<Set<RexCall>, JoinRelType>> joinConds =
+                getConditionsAndJoinType(
+                        oneSidePlan.factorIds, otherSidePlan.factorIds, multiJoin, isOuterJoin);
+        if (!joinConds.isPresent()) {
+            return Optional.empty();
+        }
+
+        Set<RexCall> conditions = joinConds.get().f0;
+        JoinRelType joinType = joinConds.get().f1;
+
+        LinkedHashSet<Integer> newFactorIds = new LinkedHashSet<>();
+        JoinPlan leftPlan;
+        JoinPlan rightPlan;
+        // put the deeper side on the left, tend to build a left-deep tree.
+        if (oneSidePlan.factorIds.size() >= otherSidePlan.factorIds.size()) {
+            leftPlan = oneSidePlan;
+            rightPlan = otherSidePlan;
+        } else {
+            leftPlan = otherSidePlan;
+            rightPlan = oneSidePlan;
+            if (isOuterJoin) {
+                joinType = (joinType == JoinRelType.LEFT) ? JoinRelType.RIGHT : JoinRelType.LEFT;
+            }
+        }
+        newFactorIds.addAll(leftPlan.factorIds);
+        newFactorIds.addAll(rightPlan.factorIds);
+
+        List<RexNode> rexCalls = new ArrayList<>(conditions);
+        Set<RexCall> newCondition =
+                convertToNewCondition(
+                        new ArrayList<>(leftPlan.factorIds),
+                        new ArrayList<>(rightPlan.factorIds),
+                        rexCalls,
+                        multiJoin);
+
+        Join newJoin =
```

Comment:
it's more safe the relBuilder is cleared before building tree

Reply:
> it's more safe the relBuilder is cleared before building tree

Done!